File: mobile/lib/models/parcel_unlock_request.dart
/// Mirrors backend/app/schemas/parcel_unlock_request.py ParcelUnlockRequestOut.
class ParcelUnlockRequest {
  final String id;
  final String orderId;
  final String? trackingNumber;
  final String riderId;
  final String? riderName;
  final String requestedById;
  final String? reason;
  final String status; // pending | approved | rejected
  final String? resolutionNote;
  final DateTime? resolvedAt;
  final DateTime? createdAt;

  ParcelUnlockRequest({
    required this.id,
    required this.orderId,
    this.trackingNumber,
    required this.riderId,
    this.riderName,
    required this.requestedById,
    this.reason,
    required this.status,
    this.resolutionNote,
    this.resolvedAt,
    this.createdAt,
  });

  bool get isPending => status == 'pending';

  factory ParcelUnlockRequest.fromJson(Map<String, dynamic> json) {
    return ParcelUnlockRequest(
      id: json['id'] as String,
      orderId: json['order_id'] as String,
      trackingNumber: json['tracking_number'] as String?,
      riderId: json['rider_id'] as String,
      riderName: json['rider_name'] as String?,
      requestedById: json['requested_by_id'] as String,
      reason: json['reason'] as String?,
      status: json['status'] as String? ?? 'pending',
      resolutionNote: json['resolution_note'] as String?,
      resolvedAt: json['resolved_at'] != null ? DateTime.tryParse(json['resolved_at'] as String) : null,
      createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at'] as String) : null,
    );
  }
}

File: backend/app/models/parcel_unlock_request.py
from sqlalchemy import Column, String, DateTime, ForeignKey, Enum
from sqlalchemy.orm import relationship

from app.database import Base
from app.models.base import TimestampMixin, UUID_TYPE, gen_uuid
from app.models.rider_status_request import RequestStatus


class ParcelUnlockRequest(Base, TimestampMixin):
    """
    A rider's request to be released from a parcel they've already claimed/
    been assigned (order.status == assigned, before pickup) - needs staff/
    admin approval before the parcel actually unassigns and goes back to the
    pool. See PATCH /staff/unlock-requests/{id}/resolve.
    """
    __tablename__ = "parcel_unlock_requests"

    id = Column(UUID_TYPE, primary_key=True, default=gen_uuid)

    order_id = Column(UUID_TYPE, ForeignKey("orders.id"), nullable=False, index=True)
    order = relationship("Order")

    rider_id = Column(UUID_TYPE, ForeignKey("riders.id"), nullable=False, index=True)
    rider = relationship("RiderProfile")

    requested_by_id = Column(UUID_TYPE, ForeignKey("users.id"), nullable=False)
    requested_by = relationship("User", foreign_keys=[requested_by_id])

    reason = Column(String(500), nullable=True)

    status = Column(Enum(RequestStatus, native_enum=False, length=20), default=RequestStatus.pending, index=True)
    resolution_note = Column(String(500), nullable=True)
    resolved_by_id = Column(UUID_TYPE, ForeignKey("users.id"), nullable=True)
    resolved_by = relationship("User", foreign_keys=[resolved_by_id])
    resolved_at = Column(DateTime(timezone=True), nullable=True)
